Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +22,72 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+
+    return totalAmphoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
+    if(current_ < 0){
+        return true;
+    }
     return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double totalHoursRemaining, msRemaining;
+        int h, m, s, ms;
+
+        if(isCharging())
+            totalHoursRemaining = totalAmpHoursUsed() / current_ * -1;
+        else
+            totalHoursRemaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / current_;
+
+        while(totalHoursRemaining > 24){
+            totalHoursRemaining -= 24;
+        }
+        //Converting totalHoursRemaining into h, m, s, ms
+        h = totalHoursRemaining;
+
+        msRemaining = (totalHoursRemaining - h) * 3600000;
+        m = msRemaining / 60000;
+        msRemaining = msRemaining - (m * 60000);
+        s = msRemaining / 1000;
+        msRemaining = msRemaining - (s * 1000);
+        ms = msRemaining;
```

Comment:
Just a rule of thumb, instead of using numbers such as 3600000 in your code, you want to place these numbers in a variable or macro to make code readable. If code has numbers without context or are not easily readable, we call them magic numbers

Reply:
https://en.wikipedia.org/wiki/Magic_number_(programming)